versiondb/README.md

```diff
@@ -0,0 +1,141 @@
+# VersionDB
+
+VersionDB stores multiple versions of on-chain state key-value pairs directly, without using a merklized tree structure like IAVL tree, both db size and query performance are much better than IAVL tree. The major lacking feature compared to IAVL tree is root hash and merkle proof generation, we still need IAVL tree for those tasks.
+
+Currently grpc query service don't support proof generation, so versiondb alone is enough to support grpc query service, there's already a `--grpc-only` flag for one to start a standalone grpc query service.
+
+There could be different implementations for the idea of versiondb, the current implementation we choose is based on rocksdb v7's experimental user-defined timestamp[^1], it'll be stored in a separate rocksdb instance, don't support other db backend yet, but the other databases in the node still support multiple backends as before.
+
+After versiondb is enabled, there's no point to keep the full archived IAVL tree anymore, it's recommended to prune the IAVL tree to keep only recent versions, for example versions within the unbonding period or even less.
+
+## Configuration
+
+To enable versiondb, add `versiondb` to the list of `store.streamers` in `app.toml` like this:
+
+```toml
+[store]
+streamers = ["versiondb"]
+```
+
+On startup, the node will create a `StreamingService` to subscribe to latest state changes in realtime and save them to versiondb, the db instance is placed at `$NODE_HOME/data/versiondb` directory, there's no way to customize the db path currently. It'll also switch grpc query service's backing store to versiondb from iavl tree, you should migrate the legacy states in advance to make the transition smooth, otherwise, the grpc queries can't see the legacy versions.
+
+If the versiondb is not empty and it's latest version doesn't match the multistore's last committed version, the startup will fail with error message `"versiondb lastest version %d doesn't match iavl latest version %d"`, that's to avoid creating gaps in versiondb accidentally. When this error happens, you just need to update versiondb to the latest version in iavl tree manually.
+
+## Migration
+
+Since our chain has pretty big now, a lot of efforts have been put to make sure the transition process can finish in practical time. The migration process will try to parallelize the tasks as much as possible, and use significant ram, but there's flags for user to control the concurrency level and ram usage to make it runnable on different machine specs.
+
+The legacy state migration process is done in two main steps:
+
+- Extract state change sets from existing archive IAVL tree.
+- Feed the change set files to versiondb.
+
+### Extract Change Sets
+
+```
+$ cronosd changeset dump --home /chain/.cronosd/ --output data distribution acc authz bank capability cronos evidence evm feegrant feeibc feemarket gov ibc mint params slashing staking transfer upgrade
+```
```

Comment:
for UX, maybe we can add `all` instead of putting these modules?

Reply:
The following comments relate to UX and can be improved later I think.

Reply:
I think it's not easy to detect all modules, because we need to include all modules since genesis, if some module is removed in previous upgrade, we can't find it in latest commit info.

Reply:
maybe just a best-effort of hardcoding it statically?

Reply:
> maybe just a best-effort of hardcoding it statically?

that could be misleading to user, if the list out dated, or there maybe difference in different stages of testnet and mainnet, it could be tedious to maintain in the future.

Reply:
I think it better maintained at the doc site, similar to how we document all the breaking upgrades and binary versions.

Reply:
https://github.com/crypto-org-chain/cronos/issues/826
Opened a separate issue to improve cmd UX later.